### PR TITLE
bpo-45152: Add HAS_CONST macro and get_const_value() function and use…

### PR DESCRIPTION
**Pull request by iritkatriel**
… them instead of explicit LOAD_CONST handling.

https://bugs.python.org/issue45152

### CODE REVIEW

File: Python/compile.c [hunk 1]
```diff
@@ -7951,6 +7951,24 @@ assemble(struct compiler *c, int addNone)
...
+    assert(HAS_CONST(opcode));
+    if (opcode == LOAD_CONST) {
+        constant = PyList_GET_ITEM(co_consts, oparg);
+    }
+
+    if (constant == NULL) {
+        PyErr_SetString(PyExc_SystemError,
+                        "Internal error: failed to get value of a constant");
+        return NULL;
+    }
```
Comment (isidentical) [inline Python/compile.c:7967]: This seems like a good case to use switch-case, especially when it gets complicated through `LOAD_NONE`, `LOAD_COMMON` etc. (with perhaps an inner switch for the latter that is depending on the oparg).

File: Python/compile.c [hunk 2]
```diff
@@ -8097,52 +8118,59 @@ optimize_basic_block(struct compiler *c, basicblock *bb, PyObject *consts)
         else {
             target = &nop;
         }
-        switch (inst->i_opcode) {
+        if (HAS_CONST(inst->i_opcode)) {
```
Comment (markshannon) [inline Python/compile.c]: Why move this out of the switch?
This code needs to understand the opcodes anyway, so adding multiple cases is fine.
```
    case LOAD_CONST:
    case LOAD_CONST_COMMON:
    case LOAD_INT:
        ...
```

**Review by markshannon — CHANGES_REQUESTED** [review]

File: Python/compile.c (hunk 2 above)
Comment (iritkatriel) [reply Python/compile.c]: We could leave it in the switch. Though it will be less obvious if you forget to add a new opcode to this case.

Comment (iritkatriel) [reply Python/compile.c]: Actually it's not true that this code needs to understand the individual load-const opcodes - this is now abstracted away in get_const_value().

Comment (markshannon) [reply Python/compile.c]: It still needs to understand all the other instructions. Handling all of them is more regular than handling almost all of them.

I prefer a switch statement here as each case is clearly labelled and I don't need to lookup what `HAS_CONST` does.

File: Python/compile.c (hunk 1 above)
Comment (iritkatriel) [reply Python/compile.c:7967]: Sure, we can turn it to a switch once we have more than one case.

**Comment by markshannon** [comment]
Thanks @iritkatriel

**Pull request merged**